adapters/spi/tomcat-adapter-spi/src/main/java/org/keycloak/adapters/tomcat/GenericPrincipalFactory.java

```diff
@@ -22,46 +22,58 @@
 
 import javax.security.auth.Subject;
 import java.security.Principal;
-import java.security.acl.Group;
 import java.util.ArrayList;
-import java.util.Collection;
 import java.util.Enumeration;
-import java.util.Iterator;
+import java.util.Collection;
 import java.util.List;
 import java.util.Set;
 
 /**
  * @author <a href="mailto:ungarida@gmail.com">Davide Ungari</a>
  * @version $Revision: 1 $
  */
-public abstract class GenericPrincipalFactory {
+public abstract class GenericPrincipalFactory implements PrincipalFactory {
 
+    @Override
     public GenericPrincipal createPrincipal(Realm realm, final Principal identity, final Set<String> roleSet) {
         Subject subject = new Subject();
         Set<Principal> principals = subject.getPrincipals();
         principals.add(identity);
-        Group[] roleSets = getRoleSets(roleSet);
-        for (int g = 0; g < roleSets.length; g++) {
-            Group group = roleSets[g];
+        final SimpleGroup[] roleSets = getRoleSets(roleSet);
+        for (SimpleGroup group : roleSets) {
             String name = group.getName();
-            Group subjectGroup = createGroup(name, principals);
+            SimpleGroup subjectGroup = createGroup(name, principals);
             // Copy the group members to the Subject group
             Enumeration<? extends Principal> members = group.members();
             while (members.hasMoreElements()) {
-                Principal role = (Principal) members.nextElement();
+                Principal role =  members.nextElement();
                 subjectGroup.addMember(role);
             }
         }
-        
-        Principal userPrincipal = getPrincipal(subject);
-        List<String> rolesAsStringList = new ArrayList<String>();
-        rolesAsStringList.addAll(roleSet);
-        GenericPrincipal principal = createPrincipal(userPrincipal, rolesAsStringList);
-        return principal;
+        return createPrincipal(getPrincipal(subject), new ArrayList<>(roleSet));
     }
 
     protected abstract GenericPrincipal createPrincipal(Principal userPrincipal, List<String> roles);
 
+    protected SimpleGroup createGroup(String name, Set<Principal> principals) {
```

Comment:
Do you mind moving this to its original location in the source file, to make the diff smaller?

(btw I'm just an interested third party, I have no relation to the Keycloak project)

Reply:
yep, I totally agree. I mean these are not my changes, but if it makes the whole thing easier to review and merge, why not...